src/hotspot/share/runtime/continuationWrapper.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2018, 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "classfile/javaClasses.inline.hpp"
+#include "jfr/jfrEvents.hpp"
+#include "jfr/support/jfrThreadId.hpp"
+#include "logging/log.hpp"
+#include "logging/logStream.hpp"
+#include "oops/oopsHierarchy.hpp"
+#include "oops/stackChunkOop.inline.hpp"
+#include "runtime/continuation.hpp"
+#include "runtime/continuationEntry.hpp"
+#include "runtime/continuationHelper.inline.hpp"
+#include "runtime/continuationWrapper.inline.hpp"
+#include "runtime/stackChunkFrameStream.inline.hpp"
+#include "runtime/thread.inline.hpp"
+
+ContinuationWrapper::ContinuationWrapper(JavaThread* thread, oop continuation)
+  : _thread(thread), _entry(thread->last_continuation()), _continuation(continuation),
+    _e_size(0), _e_num_interpreted_frames(0)
+  {
+  assert(oopDesc::is_oop(_continuation),
+         "Invalid continuation object: " INTPTR_FORMAT, p2i((void*)_continuation));
+  assert(_continuation == _entry->cont_oop(), "cont: " INTPTR_FORMAT " entry: " INTPTR_FORMAT " entry_sp: "
+         INTPTR_FORMAT, p2i((oopDesc*)_continuation), p2i((oopDesc*)_entry->cont_oop()), p2i(entrySP()));
+  disallow_safepoint();
+  read();
+}
```

Comment:
This constructor should go in the .inline file, and the intialization of the JFR fields should be #ifdefed

Reply:
I thought I should move these back.